Phydro merge into main

fabern opened the pull request:
This is the discussion thread for planning the merge of the `phydro` branch into the main model.


TODO before the merge:
- [x] publish `rsofun` documentation paper as is

TODO for the merge:
- [x] fix #220 
- [ ] fix #218 
- [ ] calibration: reduce redundancies in cost-functions and streamline two methods of 'GenSA' or 'BayesianTools'
- [ ] update `rsofun::p_model_drivers` with `readRDS(file = here::here("data/p_model_drivers_newformat.rds"))`, as well as `rsofun::p_model_drivers_vcmax25`
    - [ ] recompute `forcing`, `forcing_24h`, `forcing_daytime`, `forcing_3hrmax` for the drivers
- [ ] switch handling of `whc` from a `params_modl` to a site parameter
    - [ ] e.g. in `runread_pmodel_f.R` that sends the multi-site `data.frame` to individual simualtions?
    - [ ] or alternatively remove it from `par`: [here in R](https://github.com/geco-bern/rsofun/blob/f1366e3f2a118d6185ad48b959e50ba0d1ae4cb9/R/run_pmodel_f_bysite.R#L480-L481), [here in C](https://github.com/geco-bern/rsofun/blob/f1366e3f2a118d6185ad48b959e50ba0d1ae4cb9/src/wrappersc.c#L109-L110) and [here in F90](https://github.com/geco-bern/rsofun/blob/f1366e3f2a118d6185ad48b959e50ba0d1ae4cb9/src/sofun_r.f90#L39-L40).
- [ ] include "regression tests" into the test suites to detect alteration of results, e.g. extend current test by verifying results remain similar
- [ ] design and document a robust, user-friendly interface for the different model flags
- [ ] ...


Other suggestions:
- [ ] reduce interface complexity:
    - [ ] does `run_pmodel_f_bysite()` need to be an exported (and documented) function? Or will it anyway always be called from `runread_pmodel_f`, and we can thus keep it internal (and undocumented)?
    - [ ] remove argument `parallel` form the `runread_pmodel_f()` and `runread_biomee_f()` functions since it is equivalent with setting `ncores = 1`
    - [ ] ...

jaideep777 committed:
enable run with extra columns in driver data

jaideep777 committed:
code compiles with phydro module PROVIDED mod file is first created

jaideep777 committed:
basic working gpp calc; compiles correctly; use_phydro option

jaideep777 committed:
working gpp from instantaneous model

jaideep777 committed:
correct unit for obs LE, identified numerical issue in gs with ET_DIFFUSION

jaideep777 committed:
phydro now takes pa rather than elv as input

jaideep777 committed:
add netrad to averager, fix output units

jaideep777 committed:
jaideep's comments on equations in waterbal

jaideep777 committed:
working latent energy calcs in waterbal

jaideep777 committed:
add use_pml as program option; add pml comparison in vignette

jaideep777 committed:
feed back wcont from waterbal to phydro

jaideep777 committed:
all params are now supplied from R

jaideep777 committed:
working version with all vars in the right classes

jaideep777 committed:
basic working kphio calib with pmodel; with whc in params

jaideep777 committed:
working gensa/bayesian calibration with gpp and le

jaideep777 committed:
add full calibration in vignette

jaideep777 committed:
fix NaN in e; <--- All sites calibrated here

jaideep777 committed:
fix pmodel calibration (separate vignette)

jaideep777 committed:
param env dependencies plot

jaideep777 committed:
try calib with alpha/gamma fixed

jaideep777 committed:
Pull changes from upstream master into phydro_new

jaideep777 committed:
modify vignette to ensure that code works with pulled changes

jaideep777 committed:
get forcing_acclim dataframe into fortran

jaideep777 committed:
add acclimation forcing

jaideep777 committed:
add 2-param SWP function; limit SWP for numerical stability; bash script

jaideep777 committed:
completed pipeline for batch calibration

jaideep777 committed:
read data and output paths from cmdline args

jaideep777 committed:
update script with data/output dir inputs

jaideep777 committed:
first run for all 250 sites attempted here.

jaideep777 committed:
try P/PET to regulate le_soil; output lwp, le_soil; script to generate site list

jaideep777 committed:
improved calib script; trial with const kphio

jaideep777 committed:
add r2 and nrmse outputs

jaideep777 committed:
include obs quantities in phydro_output for convenience

jaideep777 committed:
using p50 priors

jaideep777 committed:
narrower alpha prior; r2 map

jaideep777 committed:
use new cwdx80_forcing nc file

jaideep777 committed:
Merge pull request #2 from geco-bern/master

update jaideep777/master from geco-bern/master

jaideep777 committed:
update phydro_new with changes from upstream/master

jaideep777 committed:
revert some changes to ensure phydro runs. Working for AU-ASM

jaideep777 committed:
CRITICAL BUGFIX: lines were silently deleted during merge, restore them

jaideep777 committed:
minor bugfixes, prototype multi-site calib, add new cost fn for phydro

jaideep777 committed:
export phydro cost fn

jaideep777 committed:
bugfix in cost fn phydro

jaideep777 committed:
minor bugfix

stineb committed:
Merge pull request #213 from jaideep777/phydro_new

Phydro implementation

jaideep777 committed:
trial run with phydro cost fn

jaideep777 committed:
add driver visualization in vignette; add vignette with new-format data

jaideep777 committed:
Merge branch 'hotfix/newdataformat' into phydro_new

jaideep777 committed:
updated vignettes for using pmodel and phydro with old and new data formats

jaideep777 committed:
gapfill and use forcing_acclim in phydro vignette

jaideep777 committed:
add ET formulation for pmodel (similar to phydro); move soil LE calcs to waterbal

stineb committed:
Merge pull request #216 from jaideep777/phydro_new

New driver data + ET formulation for pmodel

stineb committed:
nothing much, just for the looks

stineb committed:
implemented root zone water storage capacity-based soil moisture stress function

stineb committed:
Merge branch 'phydro' of github.com:geco-bern/rsofun into phydro

jaideep777 committed:
replace ET_* with T_* in phydro options; reserving ET for splash+phydro

jaideep777 committed:
add 3 ET options via: using_gs, using_pml

jaideep777 committed:
R interface for all ET options; update vignette to demonstrate

jaideep777 committed:
remove nested functions

jaideep777 committed:
delete unnecessary files from vignettes; use default settings with warning

jaideep777 committed:
add new format demo data

stineb committed:
Merge pull request #219 from jaideep777/phydro_new

Resolve compilation issue; 3 ET options

stineb committed:
nothing much. just niceifications.

stineb committed:
implemented PML in waterbal routine (to be used in combination with p-model). Successful compilation. Not tested.

stineb committed:
tested: identical results for gs-coupled ET with diffusion (not PML).

stineb committed:
got PML with P-model Gs running. Complemented vignette. Performs very good.

stineb committed:
nothing much

stineb committed:
bugfix to make le follow aet consistently

stineb committed:
added script for demo

fabern committed:
Update test-model-runs for PHYDRO

fabern committed:
fix warning: Unknown or uninitialised column: `forcing_acclim`

fabern committed:
Deactivate printing soilmstress

fabern committed:
test: Add consistency check for LE and AET (and soil vs canopy)

fabern committed:
fix consistency check AET_tot = AET_soil + AET_canop

fabern committed:
fix: reduce daet_soil and daet_e_soil proportionally when daet is reduced (empty bucket)

fabern committed:
Remove soilm_betao

fabern committed:
Make model parameter check more robust

fabern committed:
Make parameter calibration more robust

fabern committed:
Fix test-calibration-pmodel.R: add 'whc' and switch prior to uniform

Previously prior was changed to a TruncatedNormal, this then requires additional specifications of mean and sd for each parameter prior. For the tests it is easier to switch to a uniform prior.

fabern committed:
Fix Vcmax25 tests in test-calibration-pmodel

fabern committed:
Make non-parallel runread_pmodel_f more robust

Using rowwise() and column names instead of the positional pmap() makes it less error-prone.

fabern committed:
Make parallel runread_pmodel_f more robust and simple

fabern committed:
Make error message in required_param_names more explicit

stineb commented:
Mit diesem Merge sollten wir dann noch zuwarten. Die bisherige Strategie ist, erst das rsofun Dokumentations-Paper veröffentlichen, erst danach das hier. Das könnte aber noch dauern und wir können in Erwägung ziehen, den Merge vorzuziehen und die Dokumentation entsprechend anpassen. Allerdings wäre das eine grössere Anpassung und ich fürchte etwas, dass das mit enormem Zusatzaufwand verbunden wäre.

fabern commented:
Ja, ich denke die vorgeschlagene Strategie ist okay zuerst das paper zu veröffentlichen.

Der PR-Thread hier dient einfach als gemeinsamen Notizblock und als sichtbares Signal für den phydro branch.
Aus meiner Sicht spricht nichts dagegen, wenn dieser PR so lange im Draft Stadium bleibt wie der phydro branch.

stineb committed:
added minimal number to avoid division by zero for surface conductance when soil moisture (wcont) is zero.

stineb committed:
resolved merge conflict

fabern committed:
test: Add reference test-model runs for P and PM model

fabern committed:
test: Fix quantitative validation tests

fabern committed:
test: Skip biomee tests (TODO need to fix logp == -Inf)

fabern committed:
test: Reduce tolerance of regression tests

fabern committed:
docs: Clean comments (references)

fabern committed:
Merge branch 'phydro' into phydro-fix-le-output

fabern committed:
Merge pull request #221 from geco-bern/phydro-fix-le-output

Phydro fix le output

- `le` output (latent heat flux) was not reduced when soil was dry, even though `aet` was reduced, leading to a discrepancy. This was resolved for the total, as well as for the two constituents (soil and canopy). 
- test were updated for phydro. This involved cleaner, more robust solutions for many of the parameter checks.
- additional tests are included, checking for numerical equivalence of model outputs (a.k.a 'regression tests')
- [ ] TODO: regression test for `use_phydro == TRUE` still lacking.

fabern committed:
ci: Fix CI checks

fabern committed:
Fix vignette on sensitivity

fabern committed:
refactor: Make likelihood computation more robust

fabern committed:
fix: Vignette on sensitivity analysis

fabern committed:
feat: Parallelize Bayesian calibration of sofun

fabern committed:
test: Fix testing path for R CMD CHECK

fabern committed:
Attempt to fix R CMD Check by direct use of example driver

fabern committed:
Fix: analysis/01-sensitivity-analysis.R

fabern committed:
docs: Fix example of calib_sofun()

fabern committed:
Fix errors in R CMD Check

fabern committed:
Add dplyr:: and tidyr::

fabern committed:
Add boilerplate code for phydro regression test (currently deactivated since there is some side effects)

Commit:
Fix non initialized dpsi and psi_leaf variables.

Commit:
Merge remote-tracking branch 'origin/phydro' into phydro

fabern committed:
test: Add memory leak tests for pmodel

marcadella committed:
Merge branch 'master' into phydro

fabern committed:
feat: revert: "Parallelize Bayesian calibration of sofun"

stineb committed:
Re-designed WHC: treated as site parameter, not model parameter structurally to enable different values per site

fabern committed:
tests: Update tests for new WHC as site parameter

fabern committed:
tests: Use FR-Pue WHC in  test-model-runs.R

fabern committed:
tests: Add reference test for calibrated params

stineb committed:
made vignette a bit nicer

stineb committed:
removed stuff that was unnecessarily added to this vignette (visualisation of drivers)

stineb committed:
Merge branch 'phydro' of github.com:geco-bern/rsofun into phydro

stineb committed:
renamed vignette demonstrating pmodel calibration

stineb committed:
revamped pmodel_use vignette and complemented with demo for ET and different model flags

stineb committed:
using wind speed from forcing data frame for aerodynamic conductance

stineb committed:
scaling gs_acclim with PPFD; multiplying soil evaporation with 1.26 (Priestley-Taylor coefficient); scaling canopy conductance with LAI (calculated as -1/k*log(1-fapar) with k=0.5: Beer-Lambert coefficient for light extinction; multiplying canopy conductance with 2.0 - to be considered as a globally calibratable parameter.

stineb committed:
moved multiplication with PPFD to where gs_setpoint is calculated; introduced fixed factor 2.0 for multiplication of LAI in surface conductance calculation.

stineb committed:
nothing

stineb committed:
removed obsolete argument

Francesco-Grossi committed:
add variable use_pml in gpp functions

stineb committed:
removed argument us_pml

stineb committed:
Merge branch 'phydro' of github.com:geco-bern/rsofun into phydro

Francesco-Grossi committed:
removed argument us_pml

stineb committed:
demo and visualisations in vignette

stineb committed:
nothing really

stineb committed:
nothing really

fabern committed:
Update tests, vignette, and documentation for 'vwind' change

Previous commit (34b65317f44bc443fd636e2f70953a32b2ae7951) added 'vwind' to forcing,
to respect the testing with continuous integration this should have included updating of the
- vignettes/*.Rmd
- tests/testthat/*.R
- and documentation

Points 1 and 2 can be tested locally with devtools::build_vignettes() and with devtools::test()
Ideally, this should be enforced by CI also on branches.

fabern committed:
Update example data for 'vwind' change

Previous commit (34b65317f44bc443fd636e2f70953a32b2ae7951) added 'vwind' to forcing,
to respect the testing with continuous integration this should have included updating of the
- data/*.rda

This is now done in hindsight.

fabern committed:
Update regression tests for x2 change

Previous commit (b6ab2d43e13daf446f63c933fc1234f3acbcb56f) modified model output by multiplying surface conductance by a factor of two.
This change in model output must be reflected by updating the reference values in the regression tests.
Ideally this would be done in the same commit that alters the model structure to have a clear record of which commit changes model behavior in what way.

This is now done in hindsight.

stineb committed:
nothing really

stineb committed:
Merge branch 'phydro' of github.com:geco-bern/rsofun into phydro

fabern committed:
Activate tests for PR to phydro

fabern committed:
Update actions/upload-artifact@v4

fabern committed:
Fix arraysize to include `vwind`

fabern committed:
Enable debugging of NA on CI machine

fabern committed:
Revert "Enable debugging of NA on CI machine"

This reverts commit f0441b50093dab9e106741ed5ba157b77138324d.

fabern committed:
Merge pull request #288 from geco-bern/phydro-test-pass

Phydro: make test pass (except for 1 (issue https://github.com/geco-bern/rsofun/issues/289)) and updated documentation, vignettes and example data sets due to changes in:
- input parameters (vwind)
- model computation (factor x2 surface conductance)

fabern committed:
Activate tests for push to phydro

fabern committed:
Use single Makevars as in branches 'main' and 'luluc'

fabern committed:
Activate R-CMD-check on phydro

fabern committed:
Fix vignettes in phydro by updating dataset name

fabern committed:
Fix vignettes in phydro

fabern committed:
Fix vignettes in phydro and remove cowplot

fabern committed:
Merge branch 'phydro-test-pass-II' into phydro

Francesco-Grossi committed:
Introduce a new input parameter gw_calib (#287)

* update

* update

* update

* Homogenize whitespace

* Revert removal of data files

* Update tests, vignette, and documentation for gw change

Previous commits introduced a new input parameter gw_calib.

To respect the testing with continuous integration any change should include updating of the
- vignettes/*.Rmd
- tests/testthat/*.R
- and documentation
- data/*.rda

Ideally this would be done in the same commit as the change to the model itself.

Note that steps 1 and 2 can be tested locally with devtools::build_vignettes() and with devtools::test()

* Fix phydro bug by requiring `gw_calib`

* Use single Makevars as in branches 'main' and 'luluc'

* Remove daily and night rad output

* Remove daily and night rad output from tests

* Reduce whitespace changes by this PR

* Add gw_calib also to vignettes

* Fix gw_calib in examples of documentation

* Include gw_calib in sensitivity analysis

---------

Co-authored-by: Fabian Bernhard <10245680+fabern@users.noreply.github.com>

Commit:
changed description

stineb committed:
Bugfix: avoiding ET > soil water conent

stineb committed:
reverted hack in vignette

stineb committed:
nothing much: order of arguments, not temporary

stineb committed:
Merge branch 'phydro' of github.com:geco-bern/rsofun into phydro

stineb committed:
added line breaks to avoid truncated line

Francesco-Grossi committed:
add no vegetation option

Francesco-Grossi committed:
Fix small error

stineb committed:
nothing

stineb committed:
added runoff as output variable

stineb committed:
Merge branch 'phydro' of github.com:geco-bern/rsofun into phydro
